src/psyclone/domain/lfric/arg_ordering.py

```diff
@@ -61,22 +67,66 @@ def __init__(self, kern):
         self._generate_called = False
         self._arglist = []
 
-    def append(self, var_name):
+    def append(self, var_name, var_accesses=None, var_access_name=None,
+               var_is_array=True, mode=AccessType.READ):
         '''Appends the specified variable name to the list of all arguments.
+        If var_accesses is given, it will also record the access to the
+        variable. The name of the variable accessed can be overwritten by
+        specifying var_access_name. By default it is assumed it is an array
+        access (which can be set explicitly with var_is_array), and access
+        mode is READ (which can be set with mode).
+
         :param str var_name: the name of the variable.
+        :param var_accesses: optional class to store variable access \
+            information.
+        :type var_accesses: \
+            :py:class:`psyclone.core.access_info.VariablesAccessInfo`
+        :param str var_access_name: optional name of the variable for \
+            which access information is stored (used e.g. when the \
+            actual argument is field_proxy, but the access is to be \
+            recorded for field).
+        :param bool var_is_array: optional argument to specify if the \
+            variable access is array (defaults to True).
+        :param mode: optional access mode (defaults to READ).
+        :type mode: :py:class:`psyclone.core.access_type.AccessType`
 
         '''
         self._arglist.append(var_name)
+        if var_accesses is not None:
+            # It's an array, so add an arbitrary index value for the
+            # stored indices (which is at this stage the only way to
+            # indicate an array access).
+            if var_is_array:
+                is_array = [1]
+            else:
+                is_array = None
+
+            if var_access_name:
+                var_accesses.add_access(var_access_name, mode,
+                                        self._kern, is_array)
+            else:
+                var_accesses.add_access(var_name, mode,
+                                        self._kern, is_array)
 
-    def extend(self, list_var_name):
+    def extend(self, list_var_name, var_accesses=None):
         '''Appends all variable names in the argument list to the list of
-        all arguments.
+        all arguments. If var_accesses is given, it will also record the
+        access to the variables. Any access will be recorded as a read-only
```

Comment:
I don't believe we have symbols available. Afaik (though I might be wrong) all these implicit variables are hardcoded in various places (e.g. _declare_maps_invoke), and I think even if they are added, they will only be added when `gen_code()` is called.

Otherwise: I thought about that issue - the main issue was that `is_scalar` and `mode` would need to be arrays. That makes this method somewhat ... 'inelegant', or 'heavy'/unnecessary complicated - I was even thinking of removing it entirely (and adding a loop at the places that extend is used).
I don't care too much, but which of the following options would you prefer:
1) Add the one (see my comment about scalar/array below) or two optional parameters as lists.
2) Add the optional parameters as scalars, which will be applied to each variable in the list.
3) Remove `extend`, and add a loop (8 calls in the moved classes)

Reply:
OK, I've removed the array/scalar handling, and added a scalar parameter `mode` (that is applied to all elements of the list) - I think that's sufficient and not as 'complex'.